File: src/core/api-client.ts
import { readAuthState, ensureValidAuthState, type AuthState } from "./auth.js";
import { DEFAULT_API_BASE_URL } from "./config.js";
import { readManifest } from "./manifest.js";

// ── Types aligned with API worker (decision-record) ──

export type UserOrganization = {
  orgId: string;
  orgName: string;
  role: "reader" | "contributor" | "approver" | "admin";
};

export type ProjectSummary = {
  id: string;
  orgId: string;
  projectKey: string;
  name: string;
  isDefault: boolean;
  status: "active" | "deactivated";
  repoCount: number;
  createdAt: string;
  updatedAt: string;
};

export type UserContext = {
  user?: { id?: string; email?: string; displayName?: string };
  activeOrganization: UserOrganization | null;
  organizations: UserOrganization[];
  activeProject: ProjectSummary | null;
  projects: ProjectSummary[];
};

export type ProjectRepository = {
  repoId: string;
  installationId?: string;
  repoFullName?: string;
  defaultBranch?: string;
};

export type ProjectRepositories = {
  project: ProjectSummary;
  repositories: ProjectRepository[];
};

// Decision type aligned with API worker's nested scope+content model
export type Decision = {
  decisionId: string;
  title: string;
  type: "technical" | "product" | "business" | "governance";
  status: "proposed" | "accepted" | "deprecated" | "superseded";
  scopeType: "org" | "repo";
  scopeRepoId?: string;
  origin: "manual" | "agent";
  ownerEmail?: string;
  version: number;
  createdAt: string;
  updatedAt: string;
  createdBy?: string;
  updatedBy?: string;
};

// Full decision detail returned by GET /v1/decisions/:id
export type DecisionDetail = Decision & {
  scope?: { repos?: string[]; paths?: string[]; services?: string[]; tags?: string[] };
  content?: {
    constraints?: string[];
    supersedes?: string[];
    documentFormat?: string;
    document?: Record<string, unknown>;
    metadata?: Record<string, unknown>;
  };
  sidekick?: { suggestions?: unknown[] };
};

// Input for POST /v1/decisions — aligned with createDecisionInputSchema
export type CreateDecisionInput = {
  title: string;
  type: "technical" | "product" | "business" | "governance";
  scope?: { repos?: string[]; paths?: string[]; services?: string[]; tags?: string[] };
  content?: {
    origin?: "manual" | "agent";
    constraints?: string[];
    supersedes?: string[];
    documentFormat?: string;
    document?: Record<string, unknown>;
    metadata?: Record<string, unknown>;
  };
};

// DecisionOps workflow types — aligned with Zod schemas in @decisionrecord/shared

export type GateResult = {
  org_id: string;
  project_id: string;
  repo: string;
  branch: string;
  recordable: boolean;
  classification_reason: string;
  risk_level: "low" | "medium" | "high";
  suggested_mode: "quick" | "comprehensive";
};

export type ValidationIssue = {
  code: string;
  field: string;
  message: string;
};

export type ValidationResult = {
  valid: boolean;
  errors: ValidationIssue[];
  warnings: ValidationIssue[];
};

export type PublishResult = {
  decision_id: string;
  status: "Accepted";
  supersede_updates: Array<{ old_id: string; superseded_by: string }>;
  published_at: string;
  version: number;
};

export type DraftResult = {
  decision_id: string;
  version: number;
  status: "Proposed";
};

export type MonitoringSnapshot = Record<string, unknown>;

export type Alert = {
  id: string;
  severity: "info" | "warning" | "error";
  message: string;
  createdAt: string;
};

export type OrgConstraint = {
  id: string;
  name: string;
  description: string;
  severity: "low" | "medium" | "high";
  appliesTo: "org" | "repo" | "all";
  status: "active" | "disabled";
};

export type SearchResult = {
  decisions: Decision[];
};

export class DecisionOpsApiError extends Error {
  readonly status: number;
  constructor(status: number, message: string) {
    super(message);
    this.name = "DecisionOpsApiError";
    this.status = status;
  }
}

function toOrganization(value: Record<string, unknown>): UserOrganization {
  const role = String(value.role ?? "reader");
  return {
    orgId: String(value.orgId ?? ""),
    orgName: String(value.orgName ?? ""),
    role: role === "admin" || role === "approver" || role === "contributor" ? role : "reader",
  };
}

function toProjectSummary(value: Record<string, unknown>): ProjectSummary {
  return {
    id: String(value.id ?? ""),
    orgId: String(value.orgId ?? ""),
    projectKey: String(value.projectKey ?? ""),
    name: String(value.name ?? ""),
    isDefault: Boolean(value.isDefault),
    status: String(value.status ?? "") === "deactivated" ? "deactivated" : "active",
    repoCount: Number(value.repoCount ?? 0),
    createdAt: String(value.createdAt ?? ""),
    updatedAt: String(value.updatedAt ?? ""),
  };
}

function toProjectRepository(value: unknown): ProjectRepository {
  if (typeof value === "string") return { repoId: value };
  const obj = value as Record<string, unknown>;
  return {
    repoId: String(obj.repoId ?? obj.repo_id ?? ""),
    installationId: obj.installationId ? String(obj.installationId) : undefined,
    repoFullName: obj.repoFullName ? String(obj.repoFullName) : undefined,
    defaultBranch: obj.defaultBranch ? String(obj.defaultBranch) : undefined,
  };
}

export class DopsClient {
  private apiBaseUrl: string;
  private token: string;
  private orgId?: string;
  private projectId?: string;

  constructor(options: { apiBaseUrl?: string; token: string; orgId?: string; projectId?: string }) {
    this.apiBaseUrl = (options.apiBaseUrl ?? DEFAULT_API_BASE_URL).replace(/\/+$/, "");
    this.token = options.token;
    this.orgId = options.orgId;
    this.projectId = options.projectId;
  }

  static async fromAuth(repoPath?: string): Promise<DopsClient> {
    const auth = readAuthState();
    if (!auth) throw new Error("Not authenticated. Run: dops login");
    const validAuth = await ensureValidAuthState(auth);
    const manifest = repoPath ? readManifest(repoPath) : null;
    return new DopsClient({
      apiBaseUrl: validAuth.apiBaseUrl,
      token: validAuth.accessToken,
      orgId: manifest?.org_id ? String(manifest.org_id) : undefined,
      projectId: manifest?.project_id ? String(manifest.project_id) : undefined,
    });
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const signals: AbortSignal[] = [AbortSignal.timeout(10000)];
    let response: Response;
    try {
      response = await fetch(`${this.apiBaseUrl}${path}`, {
        method,
        headers: {
          accept: "application/json",
          authorization: `Bearer ${this.token}`,
          ...(body ? { "content-type": "application/json" } : {}),
        },
        body: body ? JSON.stringify(body) : undefined,
        signal: AbortSignal.any(signals),
      });
    } catch (error) {
      if (error instanceof DOMException && error.name === "TimeoutError") {
        throw new DecisionOpsApiError(0, `DecisionOps API timed out (${this.apiBaseUrl}).`);
      }
      const cause = error instanceof Error ? error.message : String(error);
      throw new DecisionOpsApiError(0, `Could not reach DecisionOps API: ${cause}`);
    }

    const contentType = response.headers.get("content-type") ?? "";
    const payload = contentType.includes("application/json")
      ? await response.json().catch(() => ({}))
      : await response.text().catch(() => "");

    if (!response.ok) {
      const message = typeof payload === "string"
        ? payload
        : String((payload as Record<string, unknown>).error ?? (payload as Record<string, unknown>).message ?? response.statusText);
      throw new DecisionOpsApiError(response.status, message || `Request failed (${response.status})`);
    }
    return payload as T;
  }

  // ── Auth & workspace ──

  async loadUserContext(): Promise<UserContext> {
    const payload = await this.request<Record<string, unknown>>("GET", "/v1/auth/me");
    const organizations = Array.isArray(payload.organizations)
      ? payload.organizations.map((v) => toOrganization(v as Record<string, unknown>))
      : [];
    const projects = Array.isArray(payload.projects)
      ? payload.projects.map((v) => toProjectSummary(v as Record<string, unknown>))
      : [];
    return {
      user: payload.user && typeof payload.user === "object"
        ? {
            id: String((payload.user as Record<string, unknown>).id ?? "") || undefined,
            email: String((payload.user as Record<string, unknown>).email ?? "") || undefined,
            displayName: String((payload.user as Record<string, unknown>).displayName ?? "") || undefined,
          }
        : undefined,
      activeOrganization: payload.activeOrganization && typeof payload.activeOrganization === "object"
        ? toOrganization(payload.activeOrganization as Record<string, unknown>)
        : null,
      organizations,
      activeProject: payload.activeProject && typeof payload.activeProject === "object"
        ? toProjectSummary(payload.activeProject as Record<string, unknown>)
        : null,
      projects,
    };
  }

  async loadProjectRepositories(projectId: string): Promise<ProjectRepositories> {
    const payload = await this.request<Record<string, unknown>>(
      "GET",
      `/v1/admin/projects/${encodeURIComponent(projectId)}/repositories`,
    );
    return {
      project: toProjectSummary((payload.project ?? {}) as Record<string, unknown>),
      repositories: Array.isArray(payload.repositories)
        ? payload.repositories.map(toProjectRepository)
        : [],
    };
  }

  // ── Decision CRUD (decisions-policy routes) ──

  async listDecisions(filters?: { status?: string; scopeType?: string; repoId?: string; source?: string; limit?: number }): Promise<Decision[]> {
    const params = new URLSearchParams();
    if (filters?.status) params.set("status", filters.status);
    if (filters?.scopeType) params.set("scopeType", filters.scopeType);
    if (filters?.repoId) params.set("repoId", filters.repoId);
    if (filters?.source) params.set("source", filters.source);
    if (filters?.limit) params.set("limit", String(filters.limit));
    const query = params.toString() ? `?${params.toString()}` : "";
    const payload = await this.request<Record<string, unknown>>("GET", `/v1/decisions${query}`);
    return (Array.isArray(payload.decisions) ? payload.decisions : []) as Decision[];
  }

  async getDecision(id: string): Promise<DecisionDetail> {
    const payload = await this.request<Record<string, unknown>>("GET", `/v1/decisions/${encodeURIComponent(id)}`);
    return (payload.decision ?? payload) as DecisionDetail;
  }

  async searchDecisions(query: string, filters?: { scopeType?: string; repoId?: string; status?: string; limit?: number }): Promise<SearchResult> {
    return this.request<SearchResult>("POST", "/v1/decisions/search", {
      query,
      ...filters,
    });
  }

  async createDecision(input: CreateDecisionInput): Promise<{ decision: Decision }> {
    return this.request<{ decision: Decision }>("POST", "/v1/decisions", input);
  }

  // ── DecisionOps workflow (decision-ops routes) ──

  async prepareGate(repoRef: string, taskSummary: string, changedPaths?: string[], branch?: string): Promise<GateResult> {
    return this.request<GateResult>("POST", "/v1/decision-ops/gate", {
      repo_ref: repoRef,
      task_summary: taskSummary,
      changed_paths: changedPaths,
      branch,
    });
  }

  async searchDecisionOps(terms: string[], mode: "quick" | "comprehensive" | "custom" = "quick", options?: { org_id?: string; project_id?: string; limit?: number; include_body?: boolean }): Promise<unknown> {
    return this.request<unknown>("POST", "/v1/decision-ops/search", {
      org_id: options?.org_id,
      project_id: options?.project_id,
      terms,
      mode,
      limit: options?.limit,
      include_body: options?.include_body,
    });
  }

  async createDraft(input: { org_id: string; project_id: string; title: string; context: string; decision: string; type?: string; options?: string[]; consequences?: string[]; related?: string[]; supersedes?: string[]; single_option_justification?: string; validation_plan?: { metric: string; baseline: string; target: string; by_date: string } }): Promise<DraftResult> {
    return this.request<DraftResult>("POST", "/v1/decision-ops/draft", input);
  }

  async validateDecision(input: { org_id: string; project_id: string; decision_id?: string; draft?: Record<string, unknown> }): Promise<ValidationResult> {
    return this.request<ValidationResult>("POST", "/v1/decision-ops/validate", input);
  }

  async publishDecision(input: { org_id: string; project_id: string; decision_id: string; expected_version: number }): Promise<PublishResult> {
    return this.request<PublishResult>("POST", "/v1/decision-ops/publish", input);
  }

  async getDecisionOps(id: string, projectId?: string): Promise<unknown> {
    const params = projectId ? `?project_id=${encodeURIComponent(projectId)}` : "";
    return this.request<unknown>("GET", `/v1/decision-ops/decisions/${encodeURIComponent(id)}${params}`);
  }

  // ── Monitoring (rules-monitoring routes) ──

  async getMonitoringSnapshot(): Promise<MonitoringSnapshot> {
    const payload = await this.request<Record<string, unknown>>("GET", "/v1/monitoring/snapshot");
    return (payload.snapshot ?? payload) as MonitoringSnapshot;
  }

  async getAlerts(limit = 50): Promise<Alert[]> {
    const payload = await this.request<Record<string, unknown>>("GET", `/v1/monitoring/alerts?limit=${limit}`);
    return (Array.isArray(payload.alerts) ? payload.alerts : []) as Alert[];
  }

  // ── Admin (admin routes) ──

  async listConstraints(includeDisabled = false): Promise<OrgConstraint[]> {
    const params = includeDisabled ? "?includeDisabled=true" : "";
    const payload = await this.request<Record<string, unknown>>("GET", `/v1/admin/org-constraints${params}`);
    return (Array.isArray(payload.constraints) ? payload.constraints : []) as OrgConstraint[];
  }
}

// Re-export for backwards compatibility with controlPlane consumers
export async function loadUserContext(options: { token: string; orgId?: string; apiBaseUrl?: string; signal?: AbortSignal }): Promise<UserContext> {
  const client = new DopsClient({ apiBaseUrl: options.apiBaseUrl, token: options.token, orgId: options.orgId });
  return client.loadUserContext();
}

export async function loadProjectRepositories(options: {
  token: string; orgId: string; projectId: string; apiBaseUrl?: string; signal?: AbortSignal;
}): Promise<ProjectRepositories> {
  const client = new DopsClient({ apiBaseUrl: options.apiBaseUrl, token: options.token, orgId: options.orgId });
  return client.loadProjectRepositories(options.projectId);
}

File: dops/api_client.py
from __future__ import annotations

import json
import socket
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any

from .auth import ensure_valid_auth_state, read_auth_state
from .config import DEFAULT_API_BASE_URL
from .http import HttpStatusError, default_user_agent, urlopen_with_retries
from .manifest import read_manifest
from .tls import create_ssl_context


class DecisionOpsApiError(RuntimeError):
    def __init__(self, status: int, message: str) -> None:
        super().__init__(message)
        self.status = status


def _format_auth_error_message(message: str) -> str:
    return "\n".join(
        [
            "Your saved DecisionOps login is no longer valid.",
            "Run `dops login` and try again.",
            "",
            f"Details: {message}",
        ]
    )


def _format_missing_scope_error_message(message: str) -> str:
    return "\n".join(
        [
            "Your saved DecisionOps login does not include the required permissions.",
            "Run `dops login --force` and try again.",
            "",
            f"Details: {message}",
        ]
    )


@dataclass(slots=True)
class DopsClient:
    api_base_url: str
    token: str
    org_id: str | None = None
    project_id: str | None = None
    repo_ref: str | None = None
    default_branch: str | None = None

    @classmethod
    def from_auth(cls, repo_path: str | None = None) -> "DopsClient":
        auth = read_auth_state()
        if auth is None:
            raise RuntimeError("Not authenticated. Run: dops login")
        valid_auth = ensure_valid_auth_state(auth)
        manifest = read_manifest(repo_path) if repo_path else None
        return cls(
            api_base_url=valid_auth.apiBaseUrl.rstrip("/"),
            token=valid_auth.accessToken,
            org_id=str(manifest.get("org_id")) if manifest and manifest.get("org_id") else None,
            project_id=str(manifest.get("project_id")) if manifest and manifest.get("project_id") else None,
            repo_ref=str(manifest.get("repo_ref")) if manifest and manifest.get("repo_ref") else None,
            default_branch=str(manifest.get("default_branch")) if manifest and manifest.get("default_branch") else None,
        )

    def request(self, method: str, path: str, body: Any | None = None) -> Any:
        url = f"{self.api_base_url}{path}"
        headers = {
            "accept": "application/json",
            "authorization": f"Bearer {self.token}",
            "user-agent": default_user_agent(),
        }
        payload = None
        if body is not None:
            headers["content-type"] = "application/json"
            payload = json.dumps(body).encode("utf8")
        request = urllib.request.Request(url, data=payload, method=method, headers=headers)
        try:
            response = urlopen_with_retries(request, timeout=10, context=create_ssl_context())
            content_type = response.headers.get("content-type", "")
            raw = response.body.decode("utf8")
            if "application/json" in content_type and raw:
                try:
                    return json.loads(raw)
                except json.JSONDecodeError as error:
                    raise DecisionOpsApiError(response.status, f"DecisionOps API returned invalid JSON for {url}.") from error
            return raw
        except HttpStatusError as error:
            raw = error.body.decode("utf8")
            try:
                payload = json.loads(raw) if raw else {}
            except json.JSONDecodeError:
                payload = raw
            message = (
                payload
                if isinstance(payload, str)
                else str(payload.get("error") or payload.get("message") or error.reason)
            )
            final_message = (
                _format_auth_error_message(message or str(error.reason) or f"Request failed ({error.status})")
                if error.status == 401
                else _format_missing_scope_error_message(message or str(error.reason) or f"Request failed ({error.status})")
                if error.status == 403 and "missing scope" in (message or "").lower()
                else message or f"Request failed ({error.status})"
            )
            raise DecisionOpsApiError(error.status, final_message) from error
        except socket.timeout as error:
            raise DecisionOpsApiError(0, f"DecisionOps API timed out ({self.api_base_url}).") from error
        except urllib.error.URLError as error:
            raise DecisionOpsApiError(0, f"Could not reach DecisionOps API: {error.reason}") from error

    def load_user_context(self) -> dict[str, Any]:
        payload = self.request("GET", "/v1/auth/me")
        return payload if isinstance(payload, dict) else {}

    def load_project_repositories(self, project_id: str) -> dict[str, Any]:
        return self.request("GET", f"/v1/admin/projects/{urllib.parse.quote(project_id)}/repositories")

    def attach_repository_to_project(self, project_id: str, repo_id: str) -> dict[str, Any]:
        payload = self.request(
            "POST",
            f"/v1/admin/projects/{urllib.parse.quote(project_id)}/repositories",
            {"repoId": repo_id},
        )
        return payload if isinstance(payload, dict) else {}

    def switch_active_org(self, org_id: str) -> dict[str, Any]:
        payload = self.request("POST", "/v1/auth/switch-org", {"orgId": org_id})
        return payload if isinstance(payload, dict) else {}

    def switch_active_project(self, project_id: str) -> dict[str, Any]:
        payload = self.request("POST", "/v1/auth/switch-project", {"projectId": project_id})
        return payload if isinstance(payload, dict) else {}

    def create_organization(self, name: str, *, auto_generate_service_token: bool = False) -> dict[str, Any]:
        payload = self.request(
            "POST",
            "/v1/orgs",
            {"name": name, "autoGenerateServiceToken": auto_generate_service_token},
        )
        if isinstance(payload, dict):
            organization = payload.get("organization")
            return organization if isinstance(organization, dict) else payload
        return {}

    def create_project(self, name: str, *, set_default: bool = False) -> dict[str, Any]:
        payload = self.request("POST", "/v1/projects", {"name": name, "setDefault": set_default})
        if isinstance(payload, dict):
            project = payload.get("project")
            return project if isinstance(project, dict) else payload
        return {}

    def list_decisions(self, filters: dict[str, Any] | None = None) -> list[dict[str, Any]]:
        params = urllib.parse.urlencode({key: value for key, value in (filters or {}).items() if value is not None})
        payload = self.request("GET", f"/v1/decisions{'?' + params if params else ''}")
        if isinstance(payload, dict) and isinstance(payload.get("decisions"), list):
            return payload["decisions"]
        return payload if isinstance(payload, list) else []

    def get_decision(self, decision_id: str) -> dict[str, Any]:
        payload = self.request("GET", f"/v1/decisions/{urllib.parse.quote(decision_id)}")
        if isinstance(payload, dict):
            return payload.get("decision", payload)
        return {}

    def search_decisions(self, query: str, filters: dict[str, Any] | None = None) -> dict[str, Any]:
        return self.request("POST", "/v1/decisions/search", {"query": query, **(filters or {})})

    def create_decision(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/decisions", payload)

    def prepare_gate(self, repo_ref: str, task_summary: str, changed_paths: list[str] | None = None, branch: str | None = None) -> dict[str, Any]:
        return self.request(
            "POST",
            "/v1/decision-ops/gate",
            {"repo_ref": repo_ref, "task_summary": task_summary, "changed_paths": changed_paths, "branch": branch},
        )

    def create_draft(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/decision-ops/draft", payload)

    def validate_decision(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/decision-ops/validate", payload)

    def publish_decision(self, payload: dict[str, Any]) -> dict[str, Any]:
        return self.request("POST", "/v1/decision-ops/publish", payload)

    def get_decision_ops(self, decision_id: str, project_id: str | None = None) -> dict[str, Any]:
        params = f"?project_id={urllib.parse.quote(project_id)}" if project_id else ""
        payload = self.request("GET", f"/v1/decision-ops/decisions/{urllib.parse.quote(decision_id)}{params}")
        return payload if isinstance(payload, dict) else {}

    def get_monitoring_snapshot(self) -> dict[str, Any]:
        payload = self.request("GET", "/v1/monitoring/snapshot")
        if isinstance(payload, dict):
            return payload.get("snapshot", payload)
        return {}

    def get_alerts(self, limit: int = 50) -> list[dict[str, Any]]:
        payload = self.request("GET", f"/v1/monitoring/alerts?limit={limit}")
        if isinstance(payload, dict) and isinstance(payload.get("alerts"), list):
            return payload["alerts"]
        return payload if isinstance(payload, list) else []

    def list_constraints(self, include_disabled: bool = False) -> list[dict[str, Any]]:
        params = "?includeDisabled=true" if include_disabled else ""
        payload = self.request("GET", f"/v1/admin/org-constraints{params}")
        if isinstance(payload, dict) and isinstance(payload.get("constraints"), list):
            return payload["constraints"]
        return payload if isinstance(payload, list) else []


def load_user_context(*, token: str, orgId: str | None = None, apiBaseUrl: str | None = None) -> dict[str, Any]:
    return DopsClient(api_base_url=(apiBaseUrl or DEFAULT_API_BASE_URL).rstrip("/"), token=token, org_id=orgId).load_user_context()


def load_project_repositories(*, token: str, orgId: str, projectId: str, apiBaseUrl: str | None = None) -> dict[str, Any]:
    return DopsClient(api_base_url=(apiBaseUrl or DEFAULT_API_BASE_URL).rstrip("/"), token=token, org_id=orgId).load_project_repositories(projectId)
